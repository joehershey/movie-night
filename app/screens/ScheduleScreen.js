import React, { useState } from "react";
import {
  StyleSheet,
  View,
  Text,
  Image,
  TouchableWithoutFeedback,
  SafeAreaView,
  ScrollView,
  Button,
  TextInput,
} from "react-native";

import { Card } from "react-native-elements";

import { COLORS, STYLES } from "../assets/saved";
import { FontAwesome5 } from "@expo/vector-icons";
import DateTimePicker from "@react-native-community/datetimepicker";

import TabBar from "../components/TabBar";
import TopBar from "../components/TopBar";
import Filters from "../components/Filters";
import { useLinkProps } from "@react-navigation/native";

function ScheduleScreen(props) {
  const [eventsToRender, setEvents] = useState([]);
  const [page, setPage] = useState(1);
  const [watchProviders, setWatchProviders] = useState([]);
  const [genres, setGenres] = useState([]);
  const [isLoaded, toggleLoaded] = useState(false);

  const genresKeys = [
    { name: "Action", id: 28 },
    { name: "Adventure", id: 12 },
    { name: "Animation", id: 16 },
    { name: "Comedy", id: 35 },
    { name: "Crime", id: 80 },
    { name: "Documentary", id: 99 },
    { name: "Drama", id: 18 },
    { name: "Family", id: 10751 },
    { name: "Fantasy", id: 14 },
    { name: "History", id: 36 },
    { name: "Horror", id: 27 },
    { name: "Music", id: 10402 },
    { name: "Mystery", id: 9648 },
    { name: "Romance", id: 10749 },
    { name: "Sci Fi", id: 878 },
    { name: "Thriller", id: 53 },
    { name: "War", id: 10752 },
    { name: "Western", id: 37 },
  ];

  const [name, setName] = useState("");
  const [location, setLocation] = useState("");

  // date time test
  const [date, setDate] = useState(new Date());
  const [dateString, setDateString] = useState("");
  const [mode, setMode] = useState("datetime");
  const [show, setShow] = useState(false);

  function processDate(value) {
    if (value < 10) {
      return "0" + value;
    } else {
      return value;
    }
  }

  function processTime(hours, minutes) {
    let end = "AM";
    if (minutes < 10) {
      minutes = "0" + minutes;
    }
    if (hours == 12) {
      end = "PM";
    }
    if (hours == 0) {
      hours = 12;
    }
    if (hours > 12) {
      hours = hours % 12;
      end = "PM";
    }
    return hours + ":" + minutes + " " + end;
  }

  const onChangeTimeDate = (event, selectedDate) => {
    const currentDate = selectedDate || date;
    setShow(Platform.OS === "ios");
    setDate(currentDate);
    const month = currentDate.getMonth() + 1;

    const stringDate =
      currentDate.getFullYear() +
      "-" +
      processDate(month) +
      "-" +
      processDate(currentDate.getDate()) +
      "T" +
      processDate(currentDate.getHours()) +
      ":" +
      processDate(currentDate.getMinutes()) +
      ":" +
      processDate(currentDate.getSeconds()) +
      "." +
      processDate(currentDate.getMilliseconds());
    setDateString(stringDate);
    console.log(stringDate);
    console.log(dateString);
  };

  // date time test

  const onSubmit = () => {
    const currentTime = new Date();
    // create event w api
    if (name.trim() <= 0) {
      alert("Please enter a name.");
    } else if (location.trim() <= 0) {
      alert("Please enter a location.");
    } else if (date.getTime() > currentTime.getTime) {
      alert("Please enter a date in the future.");
    } else {
      createEventAPI();
      toggleLoaded(false);
      setDate(new Date());
      setName("");
      setLocation("");
      setGenres([]);
      setWatchProviders([]);
    }
  };

  const onClear = () => {
    setDate(new Date());
    setName("");
    setLocation("");
    setGenres([]);
    setWatchProviders([]);
  };

  if (!isLoaded) {
    // works, but no events can be added yet
    getEventsAPI();
    toggleLoaded(true);
  }

  function getEventsAPI() {
    fetch(props.url + "group/" + props.group_id + "/events", {
      headers: {
        Accept: "application/json",
        "Content-Type": "application/json; charset=utf-8",
      },
      method: "GET",
    })
      .then((response) => response.json())
      .then((responseJson) => {
        // console.log("a");
        console.log(responseJson);
        setEvents(responseJson);
      })
      .catch((error) => {
        console.error(error);
      });
  }

  function createEventAPI() {
    fetch(props.url + "event/", {
      headers: {
        Accept: "application/json",
        "Content-Type": "application/json; charset=utf-8",
      },
      method: "POST",
      body: JSON.stringify({
        group_id: props.group_id,
        start_time: dateString,
        location: location,
        genre: 18,
        tmdb_movie_id: 1,
        organized_by: props.user_id,
        voting_mode: 1,
      }),
    })
      .then((response) => response.json())
      .then((responseJson) => {
        console.log(responseJson);
      })
      .catch((error) => {
        console.error(error);
        alert(error);
      });
  }

  const eventsToRenderHTML = [];
  for (const [i, event] of eventsToRender.entries()) {
    const dateTime = new Date(event.start_time);
    const month = dateTime.getMonth() + 1;

    const date =
      month + "/" + dateTime.getDate() + "/" + dateTime.getFullYear();
    const time = processTime(dateTime.getHours(), dateTime.getMinutes());

    eventsToRenderHTML.push(
      <Card key={i}>
        <Card.Title>{"Movie Night " + date}</Card.Title>
        <Card.Divider />
        <Text>{"Start Time: " + time}</Text>
        <Text>{"Location: " + event.location}</Text>
        <TouchableWithoutFeedback
          onPress={() => {
            props.setEvent(event.event_id);
            props.navigation.navigate("Event");
          }}
        >
          <View
            style={{
              margin: 10,
              padding: 5,
              backgroundColor: COLORS.primary,
              alignItems: "center",
              justifyContent: "center",
            }}
          >
            <Text
              style={{
                color: "white",
                fontSize: 20,
                alignContent: "center",
              }}
            >
              View Event
            </Text>
          </View>
        </TouchableWithoutFeedback>
      </Card>
    );
  }
  return (
    <SafeAreaView style={STYLES.container}>
      {/* Top Bar */}
      <TopBar
        navigation={props.navigation}
        screen="Landing"
        message="Groups"
      ></TopBar>
      {/* Content */}
      <View style={STYLES.content}>
        <ScrollView
          style={{ width: "100%", marginTop: 20, marginBottom: 20 }}
          contentContainerStyle={{ alignItems: "center" }}
        >
          <View style={{ width: "90%" }}>{eventsToRenderHTML}</View>

          <View>
            <View
              style={{
                backgroundColor: "lightgrey",
                padding: 5,
                alignItems: "stretch",
                marginTop: 20,
                width: 300,
              }}
            >
              <Text style={{ fontSize: 30 }}>New Movie Night</Text>

              <TextInput
                style={STYLES.formInput}
                placeholder="Name"
                onChangeText={setName}
                value={name}
              />
              <TextInput
                style={STYLES.formInput}
                placeholder={"Location (ex. 'Griffin's house')"}
                onChangeText={setLocation}
                value={location}
              />
              <View style={{}}>
                <DateTimePicker
                  testID="dateTimePicker"
                  value={date}
                  mode={mode}
                  is24Hour={true}
                  display="default"
                  onChange={onChangeTimeDate}
                />
              </View>
              <Filters
                genres={genres}
                watchProviders={watchProviders}
                setGenres={() => setGenres}
                setWatchProviders={() => setWatchProviders}
                getMovies={null}
                setPage={(a) => setPage(1)}
                inSearch={false}
              ></Filters>
            </View>
            <View style={{ flexDirection: "row", justifyContent: "center" }}>
              <TouchableWithoutFeedback onPress={onSubmit}>
                <View
                  style={{
                    borderRadius: 5,
                    margin: 10,
                    padding: 5,
                    alignItems: "center",
                    justifyContent: "center",
                    backgroundColor: COLORS.secondary,
                  }}
                >
                  <Text
                    style={{
                      color: "white",
                      fontSize: 20,
                      alignContent: "center",
                    }}
                  >
                    Submit
                  </Text>
                </View>
              </TouchableWithoutFeedback>

              <TouchableWithoutFeedback onPress={onClear}>
                <View
                  style={{
                    borderRadius: 5,
                    margin: 10,
                    padding: 5,
                    alignItems: "center",
                    justifyContent: "center",
                    backgroundColor: COLORS.primary,
                  }}
                >
                  <Text
                    style={{
                      color: "white",
                      fontSize: 20,
                      alignContent: "center",
                    }}
                  >
                    Clear Form
                  </Text>
                </View>
              </TouchableWithoutFeedback>
            </View>
          </View>
        </ScrollView>
      </View>
      {/* Tabs */}
      <TabBar navigation={props.navigation} screen="schedule"></TabBar>
    </SafeAreaView>
  );
}

export default ScheduleScreen;
