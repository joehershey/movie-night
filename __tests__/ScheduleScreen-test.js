import ScheduleScreen from '../app/screens/ScheduleScreen';
import React from 'react';
import renderer from 'react-test-renderer';

describe('Test Schedule Screen', () => {
    beforeEach( async () => {
        global.tree = renderer.create(<ScheduleScreen />).toJSON();
    });

    it("Schedule screen renders properly", async () => {
        await expect(global.tree).toMatchSnapshot();
    });
});