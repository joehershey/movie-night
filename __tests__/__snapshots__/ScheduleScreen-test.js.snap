// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Test Schedule Screen Schedule screen renders properly 1`] = `
<RCTSafeAreaView
  emulateUnlessSupported={true}
  style={
    Object {
      "backgroundColor": "#1d1f1e",
      "flex": 1,
      "flexDirection": "column",
      "justifyContent": "center",
    }
  }
>
  <View
    style={
      Object {
        "alignItems": "center",
        "backgroundColor": "#1d1f1e",
        "flex": 1,
        "flexDirection": "row",
      }
    }
  >
    <View
      accessible={true}
      focusable={true}
      onClick={[Function]}
      onResponderGrant={[Function]}
      onResponderMove={[Function]}
      onResponderRelease={[Function]}
      onResponderTerminate={[Function]}
      onResponderTerminationRequest={[Function]}
      onStartShouldSetResponder={[Function]}
      style={
        Object {
          "alignItems": "center",
          "flex": 1,
          "justifyContent": "center",
        }
      }
    >
      <Text />
      <Text
        style={
          Object {
            "color": "white",
            "margin": 5,
          }
        }
      >
        Groups
      </Text>
    </View>
    <View
      style={
        Object {
          "alignItems": "center",
          "flex": 2,
          "paddingBottom": 10,
        }
      }
    >
      <Image
        source={1}
        style={
          Object {
            "flex": 1,
            "height": 50,
            "marginBottom": 0,
            "width": 100,
          }
        }
      />
    </View>
    <View
      style={
        Object {
          "flex": 1,
        }
      }
    />
  </View>
  <View
    style={
      Object {
        "backgroundColor": "white",
        "flex": 11,
      }
    }
  >
    <RCTScrollView
      contentContainerStyle={
        Object {
          "alignItems": "center",
        }
      }
      style={
        Object {
          "marginBottom": 20,
          "marginTop": 20,
          "width": "100%",
        }
      }
    >
      <View>
        <View
          style={
            Object {
              "width": "90%",
            }
          }
        >
          <View
            style={
              Object {
                "backgroundColor": "#ffffff",
                "borderColor": "#e1e8ee",
                "borderWidth": 1,
                "margin": 15,
                "marginBottom": 0,
                "padding": 15,
                "shadowColor": "rgba(0,0,0, .2)",
                "shadowOffset": Object {
                  "height": 0,
                  "width": 0,
                },
                "shadowOpacity": 1,
                "shadowRadius": 1,
              }
            }
          >
            <View
              style={
                Object {
                  "backgroundColor": "transparent",
                }
              }
            >
              <Text
                style={
                  Object {
                    "color": "#43484d",
                    "fontSize": 18,
                    "fontWeight": "bold",
                    "marginBottom": 15,
                    "textAlign": "center",
                  }
                }
                testID="cardTitle"
              >
                Movie Night 11/5
              </Text>
              <View
                style={
                  Object {
                    "borderBottomColor": "#bcbbc1",
                    "borderBottomWidth": 0.5,
                    "marginBottom": 15,
                  }
                }
              />
              <Text>
                Start Time: 100
              </Text>
              <Text>
                Location: House 1
              </Text>
              <View
                accessible={true}
                focusable={true}
                onClick={[Function]}
                onResponderGrant={[Function]}
                onResponderMove={[Function]}
                onResponderRelease={[Function]}
                onResponderTerminate={[Function]}
                onResponderTerminationRequest={[Function]}
                onStartShouldSetResponder={[Function]}
                style={
                  Object {
                    "alignItems": "center",
                    "backgroundColor": "#1d1f1e",
                    "justifyContent": "center",
                    "margin": 10,
                    "padding": 5,
                  }
                }
              >
                <Text
                  style={
                    Object {
                      "alignContent": "center",
                      "color": "white",
                      "fontSize": 20,
                    }
                  }
                >
                  View Event
                </Text>
              </View>
            </View>
          </View>
          <View
            style={
              Object {
                "backgroundColor": "#ffffff",
                "borderColor": "#e1e8ee",
                "borderWidth": 1,
                "margin": 15,
                "marginBottom": 0,
                "padding": 15,
                "shadowColor": "rgba(0,0,0, .2)",
                "shadowOffset": Object {
                  "height": 0,
                  "width": 0,
                },
                "shadowOpacity": 1,
                "shadowRadius": 1,
              }
            }
          >
            <View
              style={
                Object {
                  "backgroundColor": "transparent",
                }
              }
            >
              <Text
                style={
                  Object {
                    "color": "#43484d",
                    "fontSize": 18,
                    "fontWeight": "bold",
                    "marginBottom": 15,
                    "textAlign": "center",
                  }
                }
                testID="cardTitle"
              >
                Movie Night 11/7
              </Text>
              <View
                style={
                  Object {
                    "borderBottomColor": "#bcbbc1",
                    "borderBottomWidth": 0.5,
                    "marginBottom": 15,
                  }
                }
              />
              <Text>
                Start Time: 101
              </Text>
              <Text>
                Location: House 3
              </Text>
              <View
                accessible={true}
                focusable={true}
                onClick={[Function]}
                onResponderGrant={[Function]}
                onResponderMove={[Function]}
                onResponderRelease={[Function]}
                onResponderTerminate={[Function]}
                onResponderTerminationRequest={[Function]}
                onStartShouldSetResponder={[Function]}
                style={
                  Object {
                    "alignItems": "center",
                    "backgroundColor": "#1d1f1e",
                    "justifyContent": "center",
                    "margin": 10,
                    "padding": 5,
                  }
                }
              >
                <Text
                  style={
                    Object {
                      "alignContent": "center",
                      "color": "white",
                      "fontSize": 20,
                    }
                  }
                >
                  View Event
                </Text>
              </View>
            </View>
          </View>
        </View>
        <View>
          <View
            style={
              Object {
                "alignItems": "stretch",
                "backgroundColor": "lightgrey",
                "marginTop": 20,
                "padding": 5,
                "width": 300,
              }
            }
          >
            <Text
              style={
                Object {
                  "fontSize": 30,
                }
              }
            >
              New Movie Night
            </Text>
            <TextInput
              allowFontScaling={true}
              placeholder="Date"
              rejectResponderTermination={true}
              style={
                Object {
                  "alignItems": "center",
                  "backgroundColor": "white",
                  "borderRadius": 10,
                  "borderWidth": 0,
                  "height": 50,
                  "justifyContent": "center",
                  "margin": 4,
                  "padding": 10,
                  "width": "90%",
                }
              }
              underlineColorAndroid="transparent"
            />
            <TextInput
              allowFontScaling={true}
              placeholder="Time"
              rejectResponderTermination={true}
              style={
                Object {
                  "alignItems": "center",
                  "backgroundColor": "white",
                  "borderRadius": 10,
                  "borderWidth": 0,
                  "height": 50,
                  "justifyContent": "center",
                  "margin": 4,
                  "padding": 10,
                  "width": "90%",
                }
              }
              underlineColorAndroid="transparent"
            />
            <TextInput
              allowFontScaling={true}
              placeholder="Genre"
              rejectResponderTermination={true}
              style={
                Object {
                  "alignItems": "center",
                  "backgroundColor": "white",
                  "borderRadius": 10,
                  "borderWidth": 0,
                  "height": 50,
                  "justifyContent": "center",
                  "margin": 4,
                  "padding": 10,
                  "width": "90%",
                }
              }
              underlineColorAndroid="transparent"
            />
            <TextInput
              allowFontScaling={true}
              placeholder="Location"
              rejectResponderTermination={true}
              style={
                Object {
                  "alignItems": "center",
                  "backgroundColor": "white",
                  "borderRadius": 10,
                  "borderWidth": 0,
                  "height": 50,
                  "justifyContent": "center",
                  "margin": 4,
                  "padding": 10,
                  "width": "90%",
                }
              }
              underlineColorAndroid="transparent"
            />
            <TextInput
              allowFontScaling={true}
              placeholder="Theme"
              rejectResponderTermination={true}
              style={
                Object {
                  "alignItems": "center",
                  "backgroundColor": "white",
                  "borderRadius": 10,
                  "borderWidth": 0,
                  "height": 50,
                  "justifyContent": "center",
                  "margin": 4,
                  "padding": 10,
                  "width": "90%",
                }
              }
              underlineColorAndroid="transparent"
            />
          </View>
          <View
            style={
              Object {
                "flexDirection": "row",
                "justifyContent": "center",
              }
            }
          >
            <View
              accessible={true}
              focusable={false}
              onClick={[Function]}
              onResponderGrant={[Function]}
              onResponderMove={[Function]}
              onResponderRelease={[Function]}
              onResponderTerminate={[Function]}
              onResponderTerminationRequest={[Function]}
              onStartShouldSetResponder={[Function]}
              style={
                Object {
                  "alignItems": "center",
                  "backgroundColor": "#166fdb",
                  "borderRadius": 5,
                  "justifyContent": "center",
                  "margin": 10,
                  "padding": 5,
                }
              }
            >
              <Text
                style={
                  Object {
                    "alignContent": "center",
                    "color": "white",
                    "fontSize": 20,
                  }
                }
              >
                Submit
              </Text>
            </View>
            <View
              accessible={true}
              focusable={false}
              onClick={[Function]}
              onResponderGrant={[Function]}
              onResponderMove={[Function]}
              onResponderRelease={[Function]}
              onResponderTerminate={[Function]}
              onResponderTerminationRequest={[Function]}
              onStartShouldSetResponder={[Function]}
              style={
                Object {
                  "alignItems": "center",
                  "backgroundColor": "#1d1f1e",
                  "borderRadius": 5,
                  "justifyContent": "center",
                  "margin": 10,
                  "padding": 5,
                }
              }
            >
              <Text
                style={
                  Object {
                    "alignContent": "center",
                    "color": "white",
                    "fontSize": 20,
                  }
                }
              >
                Clear Form
              </Text>
            </View>
          </View>
        </View>
      </View>
    </RCTScrollView>
  </View>
  <View
    style={
      Object {
        "backgroundColor": "#1d1f1e",
        "flex": 1,
        "flexDirection": "row",
        "justifyContent": "center",
        "marginTop": 10,
      }
    }
  >
    <View
      accessible={true}
      focusable={true}
      onClick={[Function]}
      onResponderGrant={[Function]}
      onResponderMove={[Function]}
      onResponderRelease={[Function]}
      onResponderTerminate={[Function]}
      onResponderTerminationRequest={[Function]}
      onStartShouldSetResponder={[Function]}
      style={
        Object {
          "alignItems": "center",
          "flex": 1,
          "justifyContent": "center",
        }
      }
    >
      <Text />
      <Text
        style={
          Object {
            "color": "#166fdb",
            "fontWeight": "bold",
            "margin": 5,
          }
        }
      >
        Schedule
      </Text>
    </View>
    <View
      accessible={true}
      focusable={true}
      onClick={[Function]}
      onResponderGrant={[Function]}
      onResponderMove={[Function]}
      onResponderRelease={[Function]}
      onResponderTerminate={[Function]}
      onResponderTerminationRequest={[Function]}
      onStartShouldSetResponder={[Function]}
      style={
        Object {
          "alignItems": "center",
          "flex": 1,
          "justifyContent": "center",
        }
      }
    >
      <Text />
      <Text
        style={
          Object {
            "color": "white",
            "fontWeight": "normal",
            "margin": 5,
          }
        }
      >
        Queue
      </Text>
    </View>
    <View
      accessible={true}
      focusable={true}
      onClick={[Function]}
      onResponderGrant={[Function]}
      onResponderMove={[Function]}
      onResponderRelease={[Function]}
      onResponderTerminate={[Function]}
      onResponderTerminationRequest={[Function]}
      onStartShouldSetResponder={[Function]}
      style={
        Object {
          "alignItems": "center",
          "flex": 1,
          "justifyContent": "center",
        }
      }
    >
      <Text />
      <Text
        style={
          Object {
            "color": "white",
            "fontWeight": "normal",
            "margin": 5,
          }
        }
      >
        Search
      </Text>
    </View>
    <View
      accessible={true}
      focusable={true}
      onClick={[Function]}
      onResponderGrant={[Function]}
      onResponderMove={[Function]}
      onResponderRelease={[Function]}
      onResponderTerminate={[Function]}
      onResponderTerminationRequest={[Function]}
      onStartShouldSetResponder={[Function]}
      style={
        Object {
          "alignItems": "center",
          "flex": 1,
          "justifyContent": "center",
        }
      }
    >
      <Text />
      <Text
        style={
          Object {
            "color": "white",
            "fontWeight": "normal",
            "margin": 5,
          }
        }
      >
        Settings
      </Text>
    </View>
  </View>
</RCTSafeAreaView>
`;
